### ISSUE TITLE
qsv dedup no longer always returns the last relevant row

### ISSUE THREAD

**Issue by tmtmtmtm** [jqnatividad/qsv#774]
version 0.22 added a test to ensure that the rows returned from `qsv dedup` would be stable, as discussed at  https://github.com/jqnatividad/qsv/discussions/80

However, this no longer seems to be true in all cases.

For example, given the `data.csv` file at https://gist.github.com/tmtmtmtm/045a8e5b9c2fd806075f631406240301, running `qsv dedup -s id` against it with versions 0.22 and 0.88 produce the different output files there.

So, for example, id Q5663401 has 4 entries in `data.csv`:

```sh
> qsv search Q5663401 data.csv 
index,id,name,lastmod
8,Q5663401,Alberto Otárola Peñaranda,2022-01-01
19,Q5663401,Alberto Otárola Peñaranda,2022-01-01
28,Q5663401,Luis Alberto Otárola Peñaranda,2022-08-26
42,Q5663401,Alberto Otárola Peñaranda,2022-12-15```
```

and under version 0.22 the last of those gets returned, as expected:

```sh
> ./qsv-22 dedup -s id data.csv | qsv search Q5663401
index,id,name,lastmod
42,Q5663401,Alberto Otárola Peñaranda,2022-12-15
```

but under 0.88 I'm now getting the second row instead:

```sh
> ./qsv-88 dedup -s id data.csv | qsv search Q5663401
index,id,name,lastmod
19,Q5663401,Alberto Otárola Peñaranda,2022-01-01
```

**Comment by tmtmtmtm** [comment]
From bisecting the releases it looks this issue was introduced in https://github.com/jqnatividad/qsv/releases/tag/0.44.0

**Comment by jqnatividad** [comment]
Thanks for the detailed report @tmtmtmtm .

I tracked down the bug to using [`par_sort_unstable_by`](https://docs.rs/rayon/latest/rayon/slice/trait.ParallelSliceMut.html#method.par_sort_unstable_by) when I parallelized `dedup` with rayon:

_Sorts the slice in parallel with a comparator function, but **might not preserve the order of equal elements.**
This sort is unstable (i.e., **may reorder equal elements**), in-place (i.e., does not allocate), and O(n * log(n)) worst-case._ (emphasis mine)

I'll change it to use `par_sort_by` instead and add your test data to `dedup` test suite.

**Comment by jqnatividad** [comment]
Hi @tmtmtmtm , can you confirm that it works as expected beyond the test data you provided?

**Comment by tmtmtmtm** [comment]
@jqnatividad Thanks for the quick fix! I'll test it out today.

(NB: the newly created release has a version number of 088.1 rather than 0.88.1, which is confusing the update check)

**Comment by jqnatividad** [comment]
Ooopss.... yanked 0.88.1 and releasing 0.88.2...

@tmtmtmtm, the binaries take about half-an-hour to compile before self-update works... so give it a try in a few minutes...

**Comment by tmtmtmtm** [comment]
@jqnatividad one issue I'm seeing with the new release is that `qsv dedup -o` seems to clobber an existing file earlier than before.

So, for example, this worked with previous releases:

```sh
qsv sort -s id data.csv | qsv dedup -o data.csv
```

But with the new release, I now get:

```
Selector name 'id' does not exist as a named header in the given CSV data.
thread 'main' panicked at 'index out of bounds: the len is 0 but the index is 0', src/cmd/dedup.rs:175:32
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

It's easy to workaround by directing the output manually, rather than via `-o`, but it might be a sign of a deeper issue.

**Comment by jqnatividad** [comment]
Hmmm.... I can't reproduce this error @tmtmtmtm ... 

Are you sure this is the erring command, because I don't see how the Selector name error would arise since you're using the column number, not the column name...

**Comment by tmtmtmtm** [comment]
@jqnatividad sorry for the bad report. I originally reported this with a `qsv sort` first, but then I edited that to avoid the red herring of sort having its own dedup built in as well. But I forgot to update the error message with that — doh!

Here's a more complete report. (I'm also happy to open this as a separate issue if you're prefer)

With `data.csv` from https://gist.github.com/tmtmtmtm/2f872e9253057f8b78d21dcfaf35a735, then running

```
qsv fmt --quote-always data.csv |  qsv dedup -o data.csv

``` 

now gives the error:

```
thread 'main' panicked at 'index out of bounds: the len is 0 but the index is 0', src/cmd/dedup.rs:175:32
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

whereas with earlier versions this works fine.

**Comment by jqnatividad** [comment]
Hi @tmtmtmtm , I'm actually now tweaking `dedup` to use an iterator instead of index access so we don't get out of bounds errors... but yeah, can you create a separate issue so we can close this one?

**Comment by tmtmtmtm** [comment]
Sure thing. Now opened as #777

**Comment by tmtmtmtm** [comment]
Other than this one issue, the latest version seems to have fixed all the other issues I was seeing, yay! 🚀

**Comment by jqnatividad** [comment]
Closing with with the creation of #777
